### ISSUE TITLE
KNN search returning less than k results when num_candidates is less than result set doc count

### ISSUE THREAD

**Issue by keanualves** [elastic/elasticsearch#115018]
### Elasticsearch Version

8.7.1

### Installed Plugins

analysis-icu

### Java Version

_bundled_

### OS Version

Linux 5.14.0-362.18.1.el9_3.x86_64

### Problem Description

I have an index containing _n_ documents with dense vectors. 
When performing an approximate knn search for the _k_ closest hits for a query vector, there are cases when Elasticsearch returns less than _k_ results even when there are more than _k_ docs in the result set containing vectors.

**Expected behaviour:**
_k_ docs should always be returned if _k_ is less than or equal to the total number of docs containing dense vectors in the result set.

**Actual behaviour:**
Less than _k_ results are returned.

### Steps to Reproduce

Unfortunately the smallest result set I could successfully reproduce this behaviour with is of size `403`, where 2 docs **don't** contain dense vectors set. Everything seems to resolve correctly when reduced further.
I've attached [bulk-docs.txt](https://github.com/user-attachments/files/17413415/bulk-docs.txt) which contains the bulk index request that can be executed to populate the index in step 2.

Although I'm running version 8.7.1, I've also been able to reproduce the same behaviour locally in version 8.15.0 using the following steps:

**1. Create index**
```
PUT /my-index
{
  "settings": {
    "number_of_shards": 1,
    "number_of_replicas": 0
  }, 
  "mappings": {
    "_source": {
      "enabled": true,
      "excludes": [
        "semantic_embedding"
      ]
    },
    "properties": {
      "@timestamp": {
        "type": "date",
        "store": true,
        "format": "epoch_millis||yyyy-MM-ddHH:mm:ss"
      },
      "semantic_embedding": {
        "type": "dense_vector",
        "dims": 128,
        "index": true,
        "similarity": "dot_product"
      }
    }
  }
}
```

**2. Insert documents**
Execute the bulk index request in [bulk-docs.txt](https://github.com/user-attachments/files/17413415/bulk-docs.txt) to insert the 403 documents.

**3. Check multiple segments**
Often during my local tests, the docs would get dispersed between 2 index segments. In these cases, it appeared to work correctly.
However, the behaviour seems to surface when all data is defined by a single segment. 
`GET _cat/segments/my-index?v=true&h=index,shard,prirep,segment,docs.count,size,searchable,compound`
If there is more than one segment, force merge into one:
`POST /my-index/_forcemerge?max_num_segments=1`

**4. Confirm 401 docs contain vectors**
```
GET my-index/_search
{
  "track_total_hits": true, 
  "query": {
    "bool": {
      "must": [
        {
          "exists": {
            "field": "semantic_embedding"
          }
        }
      ]
    }
  }
}
```

**5 Run knn search query**
```
GET my-index/_search
{
    "knn": [
        {
            "field": "semantic_embedding",
            "query_vector": [
                0.10405463725328445,
                0.06872836500406265,
                -0.03154413402080536,
                0.05876253917813301,
                0.00010786696657305583,
                -0.03682033345103264,
                -0.02495177648961544,
                -0.019133174791932106,
                0.09605824947357178,
                -0.0750531554222107,
                -0.210474893450737,
                -0.12536780536174774,
                0.04762619733810425,
                -0.029159026220440865,
                0.0062823728658258915,
                0.05450090393424034,
                -0.09959766268730164,
                0.04474235326051712,
                -0.08019277453422546,
                0.1169605404138565,
                0.046298615634441376,
                -0.11489452421665192,
                -0.006372060161083937,
                0.010990869253873825,
                0.04755079001188278,
                0.10116023570299149,
                0.000918519392143935,
                0.0027101014275103807,
                -0.1502690613269806,
                -0.14312244951725006,
                0.1215350404381752,
                0.007427239790558815,
                0.03728736564517021,
                0.1368429958820343,
                -0.11339599639177322,
                -0.11459450423717499,
                0.06264454126358032,
                0.04414265230298042,
                0.012543505989015102,
                0.02852642349898815,
                -0.12434303760528564,
                0.03353649750351906,
                0.03726150095462799,
                0.07234278321266174,
                -0.1345919668674469,
                -0.09530984610319138,
                0.1395033299922943,
                0.10010628402233124,
                -0.10837505757808685,
                0.10268478840589523,
                -0.06319449096918106,
                0.1211763396859169,
                0.03178740292787552,
                -0.01597677357494831,
                -0.06661062687635422,
                0.10101081430912018,
                0.10408773273229599,
                0.010791797190904617,
                0.039536766707897186,
                0.07304368168115616,
                -0.05732041224837303,
                0.20468732714653015,
                0.16652728617191315,
                0.07594912499189377,
                -0.013228477910161018,
                0.12920239567756653,
                -0.11352842301130295,
                -0.08272847533226013,
                -0.017150182276964188,
                -0.07550862431526184,
                -0.22037598490715027,
                0.14705835282802582,
                0.22986359894275665,
                0.00656925467774272,
                0.1398448497056961,
                -0.030111905187368393,
                0.01367101538926363,
                -0.08472618460655212,
                -0.0376223623752594,
                0.05935221537947655,
                0.03775160759687424,
                -0.08872424066066742,
                0.013909764587879181,
                0.08566707372665405,
                -0.044493574649095535,
                0.023887664079666138,
                -0.1446654200553894,
                -0.024344027042388916,
                0.16167685389518738,
                -0.03262116387486458,
                -0.12575432658195496,
                -0.0172551479190588,
                0.007975244894623756,
                -0.022601231932640076,
                -0.06843382120132446,
                0.08958141505718231,
                -0.016171058639883995,
                -0.04362731799483299,
                0.004653268028050661,
                -0.046542149037122726,
                0.0013423251220956445,
                -0.1443037986755371,
                0.0247117318212986,
                0.0070696319453418255,
                -0.008625822141766548,
                0.1437695175409317,
                0.029397638514637947,
                -0.06349257379770279,
                -0.004613952711224556,
                -0.10620557516813278,
                0.04382902756333351,
                -0.08233462274074554,
                0.09582098573446274,
                -0.15342384576797485,
                -0.04546114802360535,
                0.013855633325874805,
                -0.05375361070036888,
                0.11182045936584473,
                -0.04343185946345329,
                -0.019864404574036598,
                -0.026637574657797813,
                0.02133280225098133,
                0.08539711683988571,
                -0.014102846384048462,
                0.14996066689491272,
                -0.06198882311582565,
                -0.17123034596443176,
                -0.09001388400793076
            ],
            "k": 400,
            "num_candidates": 400
        }
    ],
    "track_total_hits": true
}
```

**Expected behaviour:**
- 400 docs returned

**Actual behaviour:**
- 384 docs returned

**Comment by carlosdelest** [comment]
@keanualves thanks for submitting this!

I've reproduced this in 8.15.0 - 8.15.3, but so far have not been able to reproduce it on the 8.16 branch.

After checking with the team, this looks like a local minima problem - we're using an HNSW graph here for retrieving approximate nearest neighbours. It's possible that not all results for the query are interconnected, so even num_candidates are searched, we don't have enough connections at the graph level.

We need to debug this in order to double check, but keep in mind that knn is approximate - we're trading off accuracy for performance. You can't expect to have the same results as in an exact knn search.

For a small number of documents after filtering (< 10k docs), most of the time you will have better performance and accuracy doing [exact knn](https://www.elastic.co/guide/en/elasticsearch/reference/current/knn-search.html#exact-knn) (see [this blog post](https://www.elastic.co/search-labs/blog/knn-exact-vs-approximate-search)).

You could experiment with graph construction parameters in [index_options](https://www.elastic.co/guide/en/elasticsearch/reference/current/dense-vector.html#dense-vector-params) (`m`, `ef_construction`, `confidence_interval`). but these are more expert mode settings that will need careful measuring for fine tuning.

**Comment by keanualves** [comment]
@carlosdelest many thanks for looking at this and the info!

My initial interpretation was that "approximate" referred to the quality of the k results (i.e. closest to the query vector) rather than the number of results that get returned. But I guess it's actually both? 

A local minima issue could explain this.

**Comment by carlosdelest** [comment]
> My initial interpretation was that "approximate" referred to the quality of the k results (i.e. closest to the query vector) rather than the number of results that get returned. But I guess it's actually both?

It can be both. Keep in mind that we're using a different data structure for retrieving the results, and the traversal can run into local minima or a lack of connections to actual nearest neighbours due to the graph construction process.

I'll try to debug this to double check. In the mean time, considering exact kNN search could be a good workaround if the filtered data size allows for that in terms of performance.

**Comment by benwtrent** [comment]
Hey @keanualves 

> My initial interpretation was that "approximate" referred to the quality of the k results (i.e. closest to the query vector) rather than the number of results that get returned. But I guess it's actually both?

We should consider the searcher getting stuck in local minima to such an extent that fewer than `num_candidate` or `k` results are returned when we know for a fact there are more results as a bug.

This will be improved in 8.16 (actually, we cannot reproduce your particular bug at all in 8.16). 

https://github.com/apache/lucene/issues/12627

https://github.com/apache/lucene/pull/13566


Ways to get around this bug in the current version is to adjust your graph building parameters as @carlosdelest suggested.

HNSW will always be approximate, but having it so disconnected that the searcher cannot progress is very frustrating.

**Comment by carlosdelest** [comment]
@keanualves , now that 8.16 is out, could you please check if your bug is resolved? We were not able to reproduce it on 8.16
